### PR TITLE
Ports TGs silicon camera expansion

### PR DESCRIPTION
**Pull request by jknpj**
neohud

The album icon shows if you're using your local(red) or AI(green) storage. The camera icon will go green when camera mode is on.

PS:. I hope i didn't forget anything when cutting this from the huge silicon update.

:cl:
 * tgs: Cyborgs can now take and upload photos to newscasters. If connected to an AI (Album HUD icon green) any photo taken by them is automatically sync'd with them. Pulsing their camera wire forces them to sync photos with their connected AI.

### CODE REVIEW

File: code/game/machinery/newscaster.dm [hunk 1]
```diff
@@ -103,35 +95,33 @@ var/list/obj/machinery/newscaster/allCasters = list() //Global list that will co
...
 	var/photo = null
 	var/channel_name = ""; //the feed channel which will be receiving the feed, or being created
-	var/c_locked=0;        //Will our new channel be locked to public submissions?
-	var/hitstaken = 0      //Death at 3 hits from an item with force>=15
+	var/c_locked = FALSE; //Will our new channel be locked to public submissions?
+	var/hitstaken = 0 //Death at 3 hits from an item with force>=15
 	var/datum/feed_channel/viewing_channel = list()
-	luminosity = 0
-	anchored = 1
+	luminosity = FALSE
```
Comment (DamianX) [inline code/game/machinery/newscaster.dm]: not a boolean

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 2]
```diff
@@ -1016,7 +1021,14 @@
 				updateicon()
 			else
 				to_chat(user, "<span class='warning'>Access denied.</span>")
-
+	else if(istype(W, /obj/item/device/toner))
+		if(toner >= tonermax)
+			to_chat(user, "The toner level of [src] is at it's highest level possible")
+		else
+			toner = CYBORG_MAX_TONER
+			user.drop_item()
```
Comment (DamianX) [inline code/modules/mob/living/silicon/robot/robot.dm]: should probably check for success

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 3]
```diff
@@ -1016,7 +1021,14 @@
...
+			qdel(W)
+			to_chat(user, "You fill the toner level of [src] to it's max capacity")
```
Comment (DamianX) [inline code/modules/mob/living/silicon/robot/robot.dm]: its vs it's, also missing a period

File: code/modules/mob/living/silicon/robot/robot.dm [hunk 4]
```diff
@@ -1016,7 +1021,14 @@
 				updateicon()
 			else
 				to_chat(user, "<span class='warning'>Access denied.</span>")
-
+	else if(istype(W, /obj/item/device/toner))
+		if(toner >= tonermax)
+			to_chat(user, "The toner level of [src] is at it's highest level possible")
```
Comment (DamianX) [inline code/modules/mob/living/silicon/robot/robot.dm]: its vs it's, also missing a period

File: code/modules/paperwork/photography.dm [hunk 5]
```diff
@@ -570,26 +610,69 @@
...
+	visible_message("[C.name] spits out a photograph from a narrow slot on it's chassis.")
+	playsound(loc, "polaroid", 75, 1, -3)
+	to_chat(C, "You print a photograph.")
+
+/obj/item/device/camera/silicon/proc/sync(var/mob/living/silicon/robot/R)
+	if(R.connected_ai && R.connected_ai.aicamera && R.aicamera) // Send images the Cyborg has taken to the AI's album upon sync.
+		if(R.aicamera.aipictures.len)
+			for(var/i in R.aicamera.aipictures)
+				if(!i in R.connected_ai.aicamera.aipictures)
+					var/datum/picture/p = i
```
Comment (DamianX) [inline code/modules/paperwork/photography.dm]: I don't think this cast is needed, same for the other loop below this

File: code/modules/paperwork/photography.dm [hunk 6]
```diff
@@ -570,26 +610,69 @@
...
+	p.pixel_x = rand(-10, 10)
+	p.pixel_y = rand(-10, 10)
+	C.toner -= CYBORG_PHOTO_COST
+	visible_message("[C.name] spits out a photograph from a narrow slot on it's chassis.")
+	playsound(loc, "polaroid", 75, 1, -3)
+	to_chat(C, "You print a photograph.")
+
+/obj/item/device/camera/silicon/proc/sync(var/mob/living/silicon/robot/R)
+	if(R.connected_ai && R.connected_ai.aicamera && R.aicamera) // Send images the Cyborg has taken to the AI's album upon sync.
+		if(R.aicamera.aipictures.len)
```
Comment (DamianX) [inline code/modules/paperwork/photography.dm]: if `len` is 0 the loop will simply not run, you can skip this

File: code/modules/paperwork/photography.dm [hunk 7]
```diff
@@ -570,26 +610,69 @@
...
+			for(var/i in R.aicamera.aipictures)
+				if(!i in R.connected_ai.aicamera.aipictures)
```
Comment (DamianX) [inline code/modules/paperwork/photography.dm]: Ackshually, can't you just rewrite the entirety of this into `R.connected_ai.aicamera.aipictures |= R.aicamera.aipictures`?

File: code/game/machinery/newscaster.dm (hunk 1 above)
Comment (jknpj) [reply code/game/machinery/newscaster.dm]: this is what i get by mass replacing 0s

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 2 above)
Comment (jknpj) [reply code/modules/mob/living/silicon/robot/robot.dm]: Good call

File: code/modules/mob/living/silicon/robot/robot.dm (hunk 3 above)
Comment (jknpj) [reply code/modules/mob/living/silicon/robot/robot.dm]: fucking TG

File: code/modules/paperwork/photography.dm (hunk 7 above)
Comment (jknpj) [reply code/modules/paperwork/photography.dm]: You're right, this is dumb.

**Pull request merged**